FreeRTOS/FreeRTOS-Kernel
# Fix prvSelectHighestPriorityTask ready list iteration

Fix prvSelectHighestPriorityTask ready list iteration. Fixes #990.

Description
-----------
Use listGET_OWNER_OF_NEXT_ENTRY to iterate through the ready task list instead of iterating from the head of the list. The SMP changes to prvSelectHighestPriorityTask added the use of vListInsertEnd to append the current task to the end of the list. vListInsertEnd, however, appends a node such that, when using listGET_OWNER_OF_NEXT_ENTRY, it is the last node returned, but it isn't inserted at the actual tail of the list. This led to a case where the list's `pxIndex` eventually managed to become the head of the list, and using `vListInsertEnd` would cause the added node to be added to the head of the list.

Test Steps
-----------
This is not very deterministic to reproduce, but I had around 7 tasks being scheduled, and many of them were sleeping for different amounts of time. The main cause seems to be the `pxIndex` of the ready lists moving to the front of the list.

Checklist:
----------

Related Issue
-----------
#990

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
